File: Software_Analysis_&_Design_F15/Projects/Part6_informed_astar_search/src/bridge/BridgeMove.java
package bridge;

import framework.Move;
import framework.State;

/**
 * This class represents moves in the Bridge Crossing problem.
 * A move object stores its move name and knows how to apply itself
 * to a bridge state to create a new state representing the result
 * of the move.
 * Note that this class extends the abstract class <b>Move</b> and
 * therefore imports <b>framework.Move</b>.
 * This class inherits the <b>getMoveName()</b> method from its parent
 * and thus it should not have an instance field for the move name.
 * Other than that, it can be essentially the same as in the previous
 * assignment.
 * @author Peter Braband
 */
public class BridgeMove extends Move {

    /**
     * Constructs a new bridge move object.
     * Note that the move name is passed to the parent constructor
     * using <b>super</b>.
     * @param moveName the name of this move.
     * It is an error if the name is not one of the following:
     * <ul>
     * <li> "P1 crosses alone" </li>
     * <li> "P2 crosses alone" </li>
     * <li> "P5 crosses alone" </li>
     * <li> "P10 crosses alone" </li>
     * <li> "P1 crosses with P2" </li>
     * <li> "P1 crosses with P5" </li>
     * <li> "P1 crosses with P10" </li>
     * <li> "P2 crosses with P5" </li>
     * <li> "P2 crosses with P10" </li>
     * <li> "P5 crosses with P10" </li>
     * </ul>
     */
    public BridgeMove(String moveName) {
        super(moveName);
        moveType = moveName;
    }
    
    /**
     * Attempts to perform this move on a given bridge state.
     * Note that this method implements the abstract <b>doMove</b> method declared
     * in the parent.
     * Thus the argument of type <b>State</b> must be cast to type
     * <b>BridgeState</b> before processing.
     * The move to perform is determined by this object's move name.
     * If the move can be performed a new bridge state object is returned that
     * reflects this move.
     * A move cannot be performed if the flashlight is not on the same side
     * as the crossing person(s), or if a pair of persons who are crossing are not
     * on the same side.
     * If the move cannot be performed <b>null</b> is returned.
     * @param otherState the bridge state on which this move is to be performed
     * @return a new bridge state reflecting the move, or <b>null</b> if it
     * cannot be performed
     */
    public State doMove(State otherState) {
        BridgeState state = (BridgeState) otherState;
        Position newP1 = state.getP1Position();
        Position newP2 = state.getP2Position();
        Position newF = state.getFlashlightPosition();
        Position newP5 = state.getP5Position();
        Position newP10 = state.getP10Position();
        Position newFlash = state.getFlashlightPosition();
        int timeToAdd = 0;
        
        
        if(moveType.contains("P1 "))
        {
            if(checkVsFlash(state.getP1Position(), state) )
            {
                newP1 = switchPos(newP1);
                timeToAdd = 1;
            }
            else
                return null;
        }
        if (moveType.contains("P2 ") || moveType.contains(" P2")){
            if(checkVsFlash(state.getP2Position(), state) )
            {
                newP2 = switchPos(newP2);
                timeToAdd = 2;
            }
            else
                return null;
        }
        if( moveType.contains("P5 ") || moveType.contains(" P5"))
        {
            if(checkVsFlash(state.getP5Position(), state))
            {
                newP5 = switchPos(newP5);
                timeToAdd = 5;
            }
            else
                return null;
        }
        if( moveType.contains("P10 ") || moveType.contains(" P10"))
        {
            if(checkVsFlash(state.getP10Position(), state))
            {
                newP10 = switchPos(newP10);
                timeToAdd = 10;
            }
            else
                return null;
        }
        
        newFlash = switchPos(newFlash);
        BridgeState newState = new BridgeState(newP1, newP2, newFlash, newP5,
                                newP10, state.getTimeSoFar() + timeToAdd);
        return newState;
    }
    
    // Private methods and instance fields should go here
    /**
     * String to store name of move for other methods to reference
     */
    private final String moveType;
    
    /**
     * Method to swap position of the variable passed in
     * Checks the position passed in and returns the opposite
     * @param old the old position to check
     * @return a Position, the opposite of that passed in
     */
    private Position switchPos(Position old){
        if(old == Position.WEST)
            return Position.EAST;
        else
            return Position.WEST;
    }
    /**
     * Method to check to see if the incoming variable is on the same side of
     * a bridge as the flashlight
     * @param Pos the position being checked against flashlight Position
     * @param state the state to check flashlight position location
     * @return a boolean value depending on if evals to true/false
     */
    private boolean checkVsFlash(Position Pos, BridgeState state){
        if(Pos == state.getFlashlightPosition() )
            return true;
        else
            return false;
    }
}

File: Software_Analysis_&_Design_F15/Labs/Lab13_Javascript/js/bridge/BridgeMove.js
var P1 = "P1 crosses alone";
var P2 = "P2 crosses alone";
var P5 = "P5 crosses alone";
var P10 = "P10 crosses alone";
var P1P2 = "P1 crosses with P2";
var P1P5 = "P1 crosses with P5";
var P1P10 = "P1 crosses with P10";
var P2P5 = "P2 crosses with P5";
var P2P10 = "P2 crosses with P10";
var P5P10 = "P5 crosses with P10";

function BridgeMove(name) {
    this.name = name;

    this.doMove = function (state) {  // Attempts to perform this move on
	var next;   
	switch (this.name) {          // Returns a new state or null.
        case P1: next = moveP1(state); break;
        case P2: next = moveP2(state); break;
        case P5: next = moveP5(state); break;
        case P10: next = moveP10(state); break;
        case P1P2: next = moveP1P2(state); break;
        case P1P5: next = moveP1P5(state); break;
        case P1P10: next = moveP1P10(state); break;
        case P2P5: next = moveP2P5(state); break;
        case P2P10: next = moveP2P10(state); break;
        case P5P10: next = moveP5P10(state); break;
        default: alert("Error in doMove for Bridge"); // Shouldn't happen!
	}
	if (next === null)
	    return null;
	return next;
    };
}

function moveP1(state) {
    
    return state.flash !== state.p1 ? null : new BridgeState(opposite(state.p1), 
			   state.p2,
                           opposite(state.flash),
			   state.p5, 
			   state.p10,
                           state.time + 1);
    
}

function moveP2(state) {
    return state.flash !== state.p2 ? null : new BridgeState(state.p1, 
			   opposite(state.p2),
                           opposite(state.flash),
			   state.p5, 
			   state.p10,
                           state.time + 2);
}

function moveP5(state) {
    return state.flash !== state.p5 ? null : new BridgeState(state.p1, 
			   state.p2,
                           opposite(state.flash),
			   opposite(state.p5), 
			   state.p10,
                           state.time + 5);
}

function moveP10(state) {  
    return state.flash !== state.p10 ? null : new BridgeState(state.p1, 
			   state.p2,
                           opposite(state.flash),
			   state.p5, 
			   opposite(state.p10),
                           state.time + 10);
}

function moveP1P2(state) { 

    if (state.p1 !== state.p2) {     
        return null;
    }

    if (state.p1 !== state.flash) {
        return null;
    }

    return new BridgeState(opposite(state.p1), 
        opposite(state.p2), 
        opposite(state.flash),
        state.p5,
        state.p10,
        state.time + 2);
}

function moveP1P5(state) {  
    if(state.p1 !== state.p5) {
        return null;
    }
    
    if(state.p1 !== state.flash) {
        return null;
    }
    
    return new BridgeState(opposite(state.p1), 
        state.p2,
        opposite(state.flash),
        opposite(state.p5), 
        state.p10,
        state.time + 5);
}

function moveP1P10(state) {  
    if(state.p1 !== state.p10) {
        return null;
    }
    
    if(state.p1 !== state.flash) {
        return null;
    }

    return new BridgeState(opposite(state.p1), 
        state.p2,
        opposite(state.flash),
        state.p5, 
        opposite(state.p10),
        state.time + 10);
}

function moveP2P5(state) {  
    if(state.p2 !== state.p5) {
        return null;
    }
    
    if(state.p2 !== state.flash) {
        return null;
    }
    
    return new BridgeState (state.p1, 
        opposite(state.p2),
        opposite(state.flash),
        opposite(state.p5), 
        state.p10,
        state.time + 5);
}

function moveP2P10(state) {  
    if(state.p2 !== state.p10) {
        return null;
    }
    
    if(state.p2 !== state.flash) {
        return null;
    }
    
    return new BridgeState (state.p1, 
        opposite(state.p2),
        opposite(state.flash),
        state.p5, 
        opposite(state.p10),
        state.time + 10);
}

function moveP5P10(state) {  
    if(state.p5 !== state.p10) {
        return null;
    }
    
    if(state.p5 !== state.flash) {
        return null;
    }
    
    return new BridgeState (state.p1, 
        state.p2,
        opposite(state.flash),
        opposite(state.p5), 
        opposite(state.p10),
        state.time + 10);
}

function opposite(side) {
    if ( side === WEST )
        return EAST;
    else
        return WEST;
}
// You must provide